torch/optim/adagrad.py

```diff
@@ -318,48 +319,50 @@ def _multi_tensor_adagrad(
     if len(params) == 0:
         return
 
-    if maximize:
-        grads = torch._foreach_neg(grads)
+    grouped_tensorlists = _group_tensors_by_device_and_dtype([params, grads, state_sums, state_steps])
+    for device_params, device_grads, device_state_sums, device_state_steps in grouped_tensorlists.values():
+
+        if maximize:
+            device_grads = torch._foreach_neg(device_grads)
+
+        device_has_sparse_grad = any(grad.is_sparse for grad in device_grads)
```

Comment:
You now unconditionally recompute this even if it was already passed in? That doesn't sound right?

Reply:
Hm, my thinking was that even if the whole batch may contain sparse grads, groups may not, so we can still use foreach to optimize these subgroups.

Is it more common for all grads to either be sparse or unsparse? If so, there may be other places I need to patch 😬

Reply:
I think it is fine if we say that we need this info. I'm just more worried about the ignored arg ;)

Reply:
I'm guessing whatever resolution we have here should be applied to https://github.com/pytorch/pytorch/pull/92338/files#r1072880613 as well?

Reply:
Yep. Can be done later.

Reply:
Yep, okay, my initial proposal is that we deprecate has_sparse_grad across the codebase.